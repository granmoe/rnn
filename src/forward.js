import Graph from './Graph'
import Mat from './Mat'
import { slidingWindow, softmax, maxIndex, sampleIndex } from './utils'

// calculates loss of model on a given sentence and returns graph to be used for backprop
// graph.runBackprop side-effects model via closures
export function computeCost({ type, model, textModel, hiddenSizes, sentence }) {
  const graph = new Graph()
  let log2ppl = 0
  let cost = 0
  let lh = {}
  const sentenceIndices = Array.from(sentence).map(c => textModel.letterToIndex[c])
  let delimitedSentence = [0, ...sentenceIndices, 0] // start and end tokens are zeros

  for (let [currentCharIndex, nextCharIndex] of slidingWindow(2, delimitedSentence)) {
    // TODO: Why "lh?" Change this...expand out to whatever the acronym stands for if possible
    lh = forwardIndex(graph, model, currentCharIndex, lh, hiddenSizes, type)
    const probs = softmax(lh.o) // compute the softmax probabilities, interpreting output as logprobs

    const nextCharProbability = probs.w[nextCharIndex]
    // binary logarithm 0 ... 1 = -Infinity ... 1
    log2ppl -= Math.log2(nextCharProbability) // accumulate binary log prob and do smoothing
    // natural logarithm, 0 ... 1 = -Infinity ... 0
    // since softmax will be between 0 and 1 exclusive, cost will always be negative
    // high prob = low cost, low prob = high cost
    cost -= Math.log(nextCharProbability)

    // write gradients into log probabilities
    lh.o.dw = probs.w
    lh.o.dw[nextCharIndex] -= 1
  }

  const perplexity = Math.pow(2, log2ppl / (sentence.length - 1))
  return { graph, perplexity, cost }
}

export function predictSentence({
  type,
  model,
  textModel,
  hiddenSizes,
  maxCharsGen = 100, // length of output
  sample = false,
  temperature = 1,
}) {
  let lh, logprobs, probs
  let graph = new Graph({ doBackprop: false }) // Just predict (forward), don't do backprop
  let sentence = ''
  let prev = {}
  let charIndex = 0

  do {
    lh = forwardIndex(graph, model, charIndex, prev, hiddenSizes, type)
    prev = lh

    logprobs = lh.o
    if (temperature !== 1 && sample) {
      // Scale log probabilities by temperature and renormalize
      // If temperature is high, logprobs will go towards zero,
      // and the softmax outputs will be more diffuse. If temperature is
      // very low, the softmax outputs will be more peaky
      logprobs.updateW(w => w / temperature)
    }
    probs = softmax(logprobs)

    charIndex = sample ? sampleIndex(probs.w) : maxIndex(probs.w)

    if (charIndex !== 0) sentence += textModel.indexToLetter[charIndex]
    // 0 index is END token (or is it the beginning of a new sentence?), maxCharsGen is a way to limit the max length of predictions
  } while (charIndex !== 0 && sentence.length <= maxCharsGen)

  return sentence
}

function forwardIndex(G, model, ix, prev, hiddenSizes, type) {
  // Could this somehow be how prev.o.dw is having an effect?
  const x = G.rowPluck(model['Wil'], ix) // char embedding for given char
  // forward prop the sequence learner
  return type === 'rnn'
    ? forwardRNN(G, model, x, prev, hiddenSizes)
    : forwardLSTM(G, model, x, prev, hiddenSizes)
}

// TODO: further refactoring here and make sure to understand everything
function forwardLSTM(graph, model, x, prev, hiddenSizes) {
  // forward prop for a single tick of LSTM
  // model contains LSTM parameters
  // x is 1D column vector with observation
  // prev is a struct containing hidden and cell
  // from previous iteration

  let hiddenPrevs, cellPrevs
  if (typeof prev.h === 'undefined') {
    hiddenPrevs = hiddenSizes.map(hiddenSize => new Mat(hiddenSize, 1))
    cellPrevs = [...hiddenPrevs]
  } else {
    hiddenPrevs = prev.h
    cellPrevs = prev.c
  }

  const { hidden, cell } = hiddenSizes.reduce(
    (result, hiddenSize, index, hiddenSizes) => {
      let inputVector = index === 0 ? x : result.hidden[index - 1]
      let hiddenPrev = hiddenPrevs[index]
      let cellPrev = cellPrevs[index]

      // input gate
      let h0 = graph.mul(model['Wix' + index], inputVector)
      let h1 = graph.mul(model['Wih' + index], hiddenPrev)
      let inputGate = graph.sigmoid(graph.add(graph.add(h0, h1), model['bi' + index]))

      // forget gate
      let h2 = graph.mul(model['Wfx' + index], inputVector)
      let h3 = graph.mul(model['Wfh' + index], hiddenPrev)
      let forgetGate = graph.sigmoid(graph.add(graph.add(h2, h3), model['bf' + index]))

      // output gate
      let h4 = graph.mul(model['Wox' + index], inputVector)
      let h5 = graph.mul(model['Woh' + index], hiddenPrev)
      let outputGate = graph.sigmoid(graph.add(graph.add(h4, h5), model['bo' + index]))

      // write operation on cells
      let h6 = graph.mul(model['Wcx' + index], inputVector)
      let h7 = graph.mul(model['Wch' + index], hiddenPrev)
      let cellWrite = graph.tanh(graph.add(graph.add(h6, h7), model['bc' + index]))

      // compute new cell activation
      let retainCell = graph.eltmul(forgetGate, cellPrev) // what do we keep from cell
      let writeCell = graph.eltmul(inputGate, cellWrite) // what do we write to cell
      let cellD = graph.add(retainCell, writeCell) // new cell contents

      // compute hidden state as gated, saturated cell activations
      let hiddenD = graph.eltmul(outputGate, graph.tanh(cellD))

      result.hidden.push(hiddenD)
      result.cell.push(cellD)
      // return [[...hidden, hiddenD], [...cell, cellD]]
      return result
    },
    { hidden: [], cell: [] },
  )

  // one decoder to outputs at end
  let output = graph.add(graph.mul(model['Whd'], hidden[hidden.length - 1]), model['bd'])

  // return cell memory, hidden representation and output
  return { h: hidden, c: cell, o: output }
}

function forwardRNN(graph, model, x, prev, hiddenSizes) {
  // forward prop for a single tick of RNN
  // model contains RNN parameters
  // x is 1D column vector with observation
  // prev is a struct containing hidden activations from last step

  // TODO: prev.h is the only thing used from prev...just pass in prev.h directly and use default value

  let hiddenPrevs
  if (typeof prev.h === 'undefined') {
    hiddenPrevs = hiddenSizes.map(hiddenSize => new Mat(hiddenSize, 1))
  } else {
    hiddenPrevs = prev.h
  }

  const h = hiddenSizes.reduce((result, hiddenSize, index) => {
    let inputVector = index === 0 ? x : result[index - 1]
    let hiddenPrev = hiddenPrevs[index]

    let h0 = graph.mul(model['Wxh' + index], inputVector)
    let h1 = graph.mul(model['Whh' + index], hiddenPrev)

    return graph.relu(graph.add(graph.add(h0, h1), model['bhh' + index]))
  }, [])

  // one decoder to outputs at end
  const o = graph.add(graph.mul(model['Whd'], h[h.length - 1]), model['bd'])

  // return cell memory, hidden representation and output
  return { h, o }
}
