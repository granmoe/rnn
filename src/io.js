import makeTrainFunc from './train'
import Mat, { RandMat, matFromJSON } from './Mat'

export function create({
  // BASIC HYPER PARAMS
  type,
  input,
  hiddenSizes = [20, 20],
  letterSize = 5,
  charCountThreshold = 1,
  // OPTIMIZATION HYPER PARAMS
  regc = 0.000001, // L2 regularization strength
  clipVal = 5, // clip gradients at this value
  decayRate = 0.9,
  smoothEps = 1e-8,
  // these are only passed in when restarting a saved model
  stepCache = {},
  totalIterations = 0,
  models = createModels({
    type,
    input,
    hiddenSizes,
    letterSize,
    charCountThreshold,
  }),
}) {
  const { model, textModel } = models

  const train = makeTrainFunc({
    type,
    hiddenSizes,
    regc,
    clipVal,
    decayRate,
    smoothEps,
    stepCache,
    totalIterations,
    model,
    textModel,
  })

  const toJSON = () =>
    JSON.stringify({
      type,
      hiddenSizes,
      regc,
      clipVal,
      totalIterations,
      models: {
        textModel,
        model: modelToJSON(model),
      },
      smoothEps,
      decayRate,
      stepCache: modelToJSON(stepCache),
    })

  return {
    train,
    toJSON,
    models: {
      model,
      textModel,
    },
    hiddenSizes,
  }
}

// TODO: This name is kind of awkward
function createModels({ type, hiddenSizes, letterSize, input, charCountThreshold }) {
  const sentences = input.split('\n').map(str => str.trim())
  const textModel = createTextModel(sentences, charCountThreshold)

  let model
  if (type === 'rnn') {
    model = initRNN(letterSize, hiddenSizes, textModel.inputSize)
  } else {
    model = initLSTM(letterSize, hiddenSizes, textModel.inputSize)
  }

  return { model, textModel }
}

function createTextModel(sentences, charCountThreshold = 1) {
  // go over all characters and keep track of all unique ones seen
  const charCounts = [...sentences.join('')].reduce((counts, char) => {
    counts[char] = counts[char] ? counts[char] + 1 : (counts[char] = 1)
    return counts
  }, {})

  // Note: inputSize is one greater than charList.length due to START and END tokens
  // START token will be index 0 in model letter vectors
  // END token will be index 0 in the next character softmax
  const initialVocabData = {
    sentences,
    letterToIndex: {},
    indexToLetter: {},
    charList: [],
    inputSize: 1,
  }

  return Object.entries(charCounts).reduce((result, [char, count]) => {
    if (count >= charCountThreshold) {
      result.charList.push(char)
      result.letterToIndex[char] = result.charList.length
      result.indexToLetter[result.charList.length] = char
      result.inputSize += 1
    }

    return result
  }, initialVocabData)
}

function initRNN(inputSize, hiddenSizes, outputSize) {
  const model = hiddenSizes.reduce((model, hiddenSize, index, hiddenSizes) => {
    const prevSize = index === 0 ? inputSize : hiddenSizes[index - 1]

    model['Wxh' + index] = new RandMat(hiddenSize, prevSize, 0.08)
    model['Whh' + index] = new RandMat(hiddenSize, hiddenSize, 0.08)
    model['bhh' + index] = new Mat(hiddenSize, 1)
  }, {})

  // decoder params
  model['Whd'] = new RandMat(outputSize, hiddenSizes[hiddenSizes.length - 1], 0.08)
  model['bd'] = new Mat(outputSize, 1)

  // letter embedding vectors
  model['Wil'] = new RandMat(outputSize, inputSize, 0, 0.08)

  return model
}

// inputSize = letterSize, outputSize = num unique chars
function initLSTM(inputSize, hiddenSizes, outputSize) {
  return hiddenSizes.reduce((model, hiddenSize, index, hiddenSizes) => {
    const prevSize = index === 0 ? inputSize : hiddenSizes[index - 1]

    // input gate
    model['Wix' + index] = new RandMat(hiddenSize, prevSize, 0.08)
    model['Wih' + index] = new RandMat(hiddenSize, hiddenSize, 0.08)
    model['bi' + index] = new Mat(hiddenSize, 1)
    // forget gate
    model['Wfx' + index] = new RandMat(hiddenSize, prevSize, 0.08)
    model['Wfh' + index] = new RandMat(hiddenSize, hiddenSize, 0.08)
    model['bf' + index] = new Mat(hiddenSize, 1)
    // output gate
    model['Wox' + index] = new RandMat(hiddenSize, prevSize, 0.08)
    model['Woh' + index] = new RandMat(hiddenSize, hiddenSize, 0.08)
    model['bo' + index] = new Mat(hiddenSize, 1)

    // cell write params
    model['Wcx' + index] = new RandMat(hiddenSize, prevSize, 0.08)
    model['Wch' + index] = new RandMat(hiddenSize, hiddenSize, 0.08)
    model['bc' + index] = new Mat(hiddenSize, 1)

    // decoder params
    model['Whd'] = new RandMat(outputSize, hiddenSize, 0.08)
    model['bd'] = new Mat(outputSize, 1)

    // letter embedding vectors
    model['Wil'] = new RandMat(outputSize, inputSize, 0.08)

    return model
  }, {})
}

export function loadFromJSON(json) {
  const args = JSON.parse(json)

  return create({
    ...args,
    stepCache: modelFromJSON(args.stepCache),
    models: {
      model: modelFromJSON(args.models.model),
      textModel: args.models.textModel,
    },
  })
}

const modelToJSON = model =>
  Object.entries(model).reduce(
    (result, [matName, mat]) => ({
      ...result,
      [matName]: mat.serialize(),
    }),
    {},
  )

const modelFromJSON = model =>
  Object.entries(model).reduce(
    (result, [matName, matJSON]) => ({
      ...result,
      [matName]: matFromJSON(matJSON),
    }),
    {},
  )
