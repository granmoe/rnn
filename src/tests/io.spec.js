import { create, loadFromJSON } from '../io'
import { predictSentence, computeCost } from '../forward'
import { rnnJSON, input } from './model-test-data'

// TODO: How to test with sample: true?
// AND: Add more variations of the two tests below
// Maybe make predictSentence and costFunc easier for caller to use, bind model args so only minimal args need to be passed, will help testing
test('predictSentence output matches snapshot', () => {
  const lstm = loadFromJSON(rnnJSON)
  const {
    models: { model, textModel },
    hiddenSizes,
  } = lstm

  const predictedSentence = predictSentence({
    model,
    textModel,
    hiddenSizes,
    sample: false,
    temperature: 1,
  })

  expect(predictedSentence).toMatchSnapshot()
})

test('costFunc output matches snapshot', () => {
  const lstm = loadFromJSON(rnnJSON)
  const {
    models: { model, textModel },
    hiddenSizes,
  } = lstm

  const cost = computeCost({
    model,
    textModel,
    hiddenSizes,
    sentence: 'this is a test sentence',
  })

  expect(cost).toMatchSnapshot()
})

// TODO: Not sure how to test this...only thing I can do is call any pure funcs (like predictSentence and costFunc tests above)
test('loadFromJSON produces same result as create', () => {
  const lstmModel = create({
    type: 'lstm',
    input,
    letterSize: 5,
    hiddenSizes: [20, 20],
  })

  lstmModel.train()
  const lstmModelJSON = lstmModel.toJSON()
  const rehydratedModel = loadFromJSON(lstmModelJSON)

  expect(lstmModelJSON).toEqual(rehydratedModel.toJSON())
})
