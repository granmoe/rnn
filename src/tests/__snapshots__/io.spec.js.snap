// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`costFunc output matches snapshot 1`] = `
Object {
  "cost": 53.44277632796922,
  "graph": Graph {
    "backpropFuncs": Array [
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
    ],
    "doBackprop": true,
  },
  "perplexity": 11.349992757446673,
}
`;

exports[`predictSentence output matches snapshot 1`] = `"the stong the stong"`;
